### PR TITLE
Feat: Make OData v4 docu release ready

### PR DESCRIPTION
**Pull request by FrankEssenberger**
## Context

What I have done:
- Removed all the experimental flags
- Added a few more types and interfaces

Question:
Do I have to do something about the /v4 export? If I removed it and added the /odata/v4 next to the /odata/v2 export in the root index.ts, I got of course many compile errors that there are a lot of double exports? I can of course fix this by either not exporting these methods anymore or renaming them, but to we want to do that?

## Definition of Done

Please consider all items and remove only if not applicable.

- [x] Tests created/adjusted for your changes.
- [x] Release notes updated.
  * Provide sufficient context so that each entry can be understood on its own.
  * Be specific about names of functions, classes, modules, etc.
  * Describe when or where this is relevant
  * Use indicative and present tense. For example, write "Provide function `name` that does X in order to Y" over "Now X can be done by calling a new function".
- [x] PR title adheres to [conventional commit guidelines](https://www.conventionalcommits.org) (please note that only `fix:` and `feat:` will end up in the release notes)
- [x] If applicable: Properly documented (JSDoc of public API)
- [x] If applicable: Check if `node run doc` still works.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 - Include fallback `Edm.Any` for previously unsupported edm types like Edm.Geography.
 Type safety and related features like URI conversion are not supported for the fallback type.  
+- OData v4 is moven from the experimental state to GA.
```
Comment (jjtang1985) [inline CHANGELOG.md:23]:
```suggestion
- OData v4 is moved from the experimental state to GA.
```

**Comment by FrankEssenberger** [comment]
@artemkovalyov would be great if you find time to go over the docs. I am pretty sure I have typos and so on.

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,7 @@
 
 - Include fallback `Edm.Any` for previously unsupported edm types like Edm.Geography.
 Type safety and related features like URI conversion are not supported for the fallback type.  
+- OData v4 is moved from the experimental state to GA.
 
 ## Improvements
```
Comment (artemkovalyov) [inline CHANGELOG.md:25]:
```suggestion
- Include fallback `Edm.Any` for previously unsupported EDM types like Edm.Geography.
Type safety and related features like URI conversion are not supported for the fallback type.  
- OData v4 is moved from the experimental state to GA.

## Improvements
```

File: packages/core/src/odata/common/entity-deserializer.ts [hunk 3]
```diff
@@ -18,22 +18,53 @@ import {
...
- * @experimental This is experimental and is subject to change. Use with caution.
+ * Interface representing the return type of the builder function [[entityDeserializer]]
  */
-export function entityDeserializer(
-  edmToTs,
-  extractODataETag,
-  extractDataFromOneToManyLink: (arg) => any[] | undefined
-) {
+export interface EntityDeserializer<EntityT extends EntityBase = any> {
+  // TODO: extractCustomFields should not be exported here. This was probably done only for testing
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/entity-deserializer.ts:32]: Should we create a BLI for this and remove the comment to get release ready?

File: packages/core/src/odata/common/entity-deserializer.ts [hunk 4]
```diff
@@ -18,22 +18,53 @@ import {
...
+type ExtractDataFromOneToManyLinkType = (data: any) => any[] | undefined;
+
+/**
+ * Constructs a entityDeserializer given the OData v2 or v4 specific methods.
+ * The concrete deserializers are created in odata/v2/entity-deserializer.ts and odata/v4/entity-deserializer.ts
+ * @param edmToTs - Converters  emd input to ts values.
+ * @param extractODataETag - Extractor for the Etag.
+ * @param extractDataFromOneToManyLink - Extractor for data related to one to many links.
+ * @returns a entity deserializer as defined by [[EntityDeserializer]]
+ */
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/entity-deserializer.ts:62]:
```suggestion
/**
 * Constructs an entityDeserializer given the OData v2 or v4 specific methods.
 * The concrete deserializers are created in odata/v2/entity-deserializer.ts and odata/v4/entity-deserializer.ts
 * @param edmToTs - Converters  emd input to ts values.
 * @param extractODataETag - Extractor for the Etag.
 * @param extractDataFromOneToManyLink - Extractor for data related to one to many links.
 * @returns a entity deserializer as defined by [[EntityDeserializer]]
 */
```

File: packages/core/src/odata/common/entity-deserializer.ts [hunk 5]
```diff
@@ -18,22 +18,53 @@ import {
...
+export function entityDeserializer<EdmT extends EdmType, EntityT, JsonT>(
+  edmToTs: EdmToTsType<EdmT>,
+  extractODataETag: ExtractODataETagType,
+  extractDataFromOneToManyLink: ExtractDataFromOneToManyLinkType
+): EntityDeserializer {
   /**
    * Extracts all custom fields from the JSON payload for a single entity.
    * In this context, a custom fields is every property that is not known in the corresponding entity class.
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/entity-deserializer.ts:70]:
```suggestion
  /**
   * Extracts all custom fields from the JSON payload for a single entity.
   * In this context, a custom field is every property that is not known in the corresponding entity class.
```

File: packages/core/src/odata/common/entity-serializer.ts [hunk 6]
```diff
@@ -22,11 +22,39 @@ const logger = createLogger({
...
+  value: any,
+  edmType: EdmTypeShared<'v2'> | EdmTypeShared<'v4'>
+) => any;
+
+/**
+ * Constructs a entitySerializer given the OData v2 or v4 specific tsToEdm method.
+ * The concrete serializers are created in odata/v2/entity-serializer.ts and odata/v4/entity-serializer.ts
+ * @param tsToEdm - Converters ts input to edm values
+ * @returns a entity serializer as defined by [[EntitySerializer]]
+ */
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/entity-serializer.ts:56]:
```suggestion
/**
 * Constructs an entitySerializer given the OData v2 or v4 specific tsToEdm method.
 * The concrete serializers are created in odata/v2/entity-serializer.ts and odata/v4/entity-serializer.ts
 * @param tsToEdm - Converters ts input to edm values
 * @returns a entity serializer as defined by [[EntitySerializer]]
 */
```

File: packages/core/src/odata/common/uri-conversion/get-filter.ts [hunk 7]
```diff
@@ -13,14 +13,29 @@ import {
...
+  getFilter: GetFilterType<EntityT>;
+}
+
 /**
- * @experimental This is experimental and is subject to change. Use with caution.
+ * Creates a getFilter function using the OData v2 or OData v4 uri converter.
+ * The concrete filter getters are initiated in odata/v2/uri-conversion/odata-uri.ts and odata/v4/uri-conversion/odata-uri.ts.
+ *
+ * @param uriConverter Uri converter for v2 or v4.
+ * @returns The filter getter. See interface [[GetFilter]]
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/uri-conversion/get-filter.ts:36]:
```suggestion
 * Creates a getFilter function using the OData v2 or OData v4 URI converter.
 * The concrete filter getters are initiated in odata/v2/uri-conversion/odata-uri.ts and odata/v4/uri-conversion/odata-uri.ts.
 *
 * @param uriConverter Uri converter for v2 or v4.
 * @returns The filter getter. See interface [[GetFilter]]
```

File: packages/core/src/odata/common/uri-conversion/get-resource-path.ts [hunk 8]
```diff
@@ -4,18 +4,32 @@ import { createLogger, MapType } from '@sap-cloud-sdk/util';
...
+
+interface GetResourcePathForKeys<EntityT extends EntityBase = any> {
+  getResourcePathForKeys: GetResourcePathForKeysType<EntityT>;
+}
+
 /**
- * @experimental This is experimental and is subject to change. Use with caution.
+ * Creates a getResourcePathForKeys function using the OData v2 or OData v4 uri converter.
+ * The concrete instances for v2 or v4 are initiated in odata/v2/uri-conversion/odata-uri.ts and odata/v4/uri-conversion/odata-uri.ts.
+ *
```
Comment (artemkovalyov) [inline packages/core/src/odata/common/uri-conversion/get-resource-path.ts:26]:
```suggestion
 * Creates a getResourcePathForKeys function using the OData v2 or OData v4 URI converter.
 * The concrete instances for v2 or v4 are initiated in odata/v2/uri-conversion/odata-uri.ts and odata/v4/uri-conversion/odata-uri.ts.
 *
```

File: packages/core/src/odata/v2/request-builder/get-all-request-builder.ts [hunk 9]
```diff
@@ -21,8 +21,10 @@ import { oDataUri } from '../uri-conversion';
 import { getCollectionResult } from './response-data-accessor';
 
 /**
- * Create OData request to get multiple entities based on the configuration of the request. A `GetAllRequestBuilder` allows to restrict the response in multiple dimensions.
+ * Create OData request to get multiple entities based on the configuration of the request.
+ * A `GetAllRequestBuilder` allows to restrict the response in multiple dimensions.
```
Comment (artemkovalyov) [inline packages/core/src/odata/v2/request-builder/get-all-request-builder.ts:25]:
```suggestion
 * Create an OData request to get multiple entities based on the configuration of the request.
 * A `GetAllRequestBuilder` allows restricting the response in multiple dimensions.
```

File: packages/core/src/odata/v2/request-builder/get-by-key-request-builder.ts [hunk 10]
```diff
@@ -19,8 +19,9 @@ import { ODataGetByKeyRequestConfig } from '../../common/request/odata-get-by-ke
 import { oDataUri } from '../uri-conversion';
 import { getSingleResult } from './response-data-accessor';
 /**
- * Create OData request to get a single entity based on its key properties. A `GetByKeyRequestBuilder` allows to restrict the response to a selection of fields,
- * where no selection is equal to selecting all fields.
+ * Create OData request to get a single entity based on its key properties.
+ * The properties available in the response can be restricted by creating a [[GetAllRequestBuilder.select selection]], where no selection is equal to selecting all fields.
+ * Note that navigational properties are automatically expanded if they included in a  select.
```
Comment (artemkovalyov) [inline packages/core/src/odata/v2/request-builder/get-by-key-request-builder.ts:24]:
```suggestion
 * Create an OData request to get a single entity based on its key properties.
 * The properties available in the response can be restricted by creating a [[GetAllRequestBuilder.select selection]], where no selection is equal to selecting all fields.
 * Note that navigational properties are automatically expanded if they included in a  select.
```

File: packages/core/src/odata/v2/request-builder/response-data-accessor.ts [hunk 11]
```diff
@@ -7,11 +7,26 @@ const logger = createLogger({
...
+/**
+ * Methods to extract the data from OData v2 responses.
+ */
+
+/**
+ * Extract the collection data from the repsonse.
+ * If the data does not contain a collection an empty array is returned.
+ * @param data - Response of the OData v2 service
+ * @returns any[] - Collection extracted from the repsponse
+ */
```
Comment (artemkovalyov) [inline packages/core/src/odata/v2/request-builder/response-data-accessor.ts:19]:
```suggestion
/**
 * Methods to extract the data from OData v2 responses.
 */

/**
 * Extract the collection data from the response.
 * If the data does not contain a collection an empty array is returned.
 * @param data - Response of the OData v2 service
 * @returns any[] - Collection extracted from the response
 */
```

File: packages/core/src/odata/v2/uri-conversion/get-expand.ts [hunk 12]
```diff
@@ -3,6 +3,15 @@
...
 
+/**
+ * Get an object containing the given expand as query parameter, or an empty object if none was given.
+ * In this OData v2 expand, selected properties are automatically added to the expand.
+ *
+ * @typeparam EntityT - Type of the entity to expand on
+ * @param selects - The selects which are expanded if necessary
+ * @param entityConstructor - Constructor type of the entity to expand on
+ * @returns An object containing the query parameter or an empty object
+ */
```
Comment (artemkovalyov) [inline packages/core/src/odata/v2/uri-conversion/get-expand.ts:14]:
```suggestion
/**
 * Get an object containing the given expand as a query parameter, or an empty object if none was given.
 * In this OData v2 expand, selected properties are automatically added to the expand.
 *
 * @typeparam EntityT - Type of the entity to expand on
 * @param selects - The selects which are expanded if necessary
 * @param entityConstructor - Constructor type of the entity to expand on
 * @returns An object containing the query parameter or an empty object
 */
```

File: packages/core/src/odata/v4/request-builder/action-import-request-builder.ts [hunk 13]
```diff
@@ -7,9 +7,9 @@ import {
 import { ActionFunctionImportRequestBuilderBase } from '../../common/request-builder/action-function-import-request-builder-base';
 
 /**
- * Create OData request to execute a action import.
- * @typeparam ParametersT - Type of the function import parameters
- * @typeparam ReturnT - Type of the function import return value
+ * Create OData request to execute an action import.
+ * @typeparam ParametersT - Type of the action import parameters
+ * @typeparam ReturnT - Type of the action import return value
```
Comment (artemkovalyov) [inline packages/core/src/odata/v4/request-builder/action-import-request-builder.ts:12]:
```suggestion
 * Create an OData request to execute an action import.
 * @typeparam ParametersT - Type of the action import parameters
 * @typeparam ReturnT - Type of the action import return value
```

File: packages/core/src/odata/v4/request-builder/get-all-request-builder.ts [hunk 14]
```diff
@@ -24,8 +24,10 @@ import { OneToManyLink } from '../../common/selectable/one-to-many-link';
 import { getCollectionResult } from './response-data-accessor';
 
 /**
- * Create OData request to get multiple entities based on the configuration of the request. A `GetAllRequestBuilder` allows to restrict the response in multiple dimensions.
- * The properties available in the response can be restricted by creating a [[GetAllRequestBuilder.select selection]], where no selection is equal to selecting all fields.
+ * Create OData request to get multiple entities based on the configuration of the request.
+ * A `GetAllRequestBuilder` allows to restrict the response in multiple dimensions.
+ * The properties available in the response can be restricted by creating a [[GetAllRequestBuilder.select selection]], where no selection is equal to selecting all fields of the entity.
+ * Navigational properties need to expanded explicitly by [[GetAllRequestBuilder.expand]].
```
Comment (artemkovalyov) [inline packages/core/src/odata/v4/request-builder/get-all-request-builder.ts:30]:
```suggestion
 * Create an OData request to get multiple entities based on the configuration of the request.
 * A `GetAllRequestBuilder` allows restricting the response in multiple dimensions.
 * The properties available in the response can be restricted by creating a [[GetAllRequestBuilder.select selection]], where no selection is equal to selecting all fields of the entity.
 * Navigational properties need to expanded explicitly by [[GetAllRequestBuilder.expand]].
```

File: packages/core/src/odata/v4/request-builder/get-by-key-request-builder.ts [hunk 15]
```diff
@@ -21,7 +21,9 @@ import { oDataUri } from '../uri-conversion';
 import { HttpReponse } from '../../../http-client';
 import { getSingleResult } from './response-data-accessor';
 /**
- * Create OData request to get a single entity based on its key properties. A `GetByKeyRequestBuilder` allows to restrict the response to a selection of fields,
+ * Create OData request to get a single entity based on its key properties.
+ * The properties available in the response can be restricted by creating a [[GetByKeyRequestBuilder.select selection]], where no selection is equal to selecting all fields of the entity.
+ * Navigational properties need to expanded explicitly by [[GetAllRequestBuilder.expand]].
```
Comment (artemkovalyov) [inline packages/core/src/odata/v4/request-builder/get-by-key-request-builder.ts:26]:
```suggestion
 * Create an OData request to get a single entity based on its key properties.
 * The properties available in the response can be restricted by creating a [[GetByKeyRequestBuilder.select selection]], where no selection is equal to selecting all fields of the entity.
 * Navigational properties need to expanded explicitly by [[GetAllRequestBuilder.expand]].
```

File: packages/core/src/odata/v4/uri-conversion/get-expand.ts [hunk 16]
```diff
@@ -2,15 +2,29 @@
...
 }
 
+/**
+ * Get an object containing the given expand as query parameter, or an empty object if none was given.
+ *
+ * @typeparam EntityT - Type of the entity to expand on
+ * @param expands - The expands to transform to a query parameter
+ * @param entityConstructor - Constructor type of the entity to expand on
+ * @returns An object containing the query parameter or an empty object
+ */
```
Comment (artemkovalyov) [inline packages/core/src/odata/v4/uri-conversion/get-expand.ts:27]:
```suggestion
/**
 * Get an object containing the given expand as a query parameter, or an empty object if none was given.
 *
 * @typeparam EntityT - Type of the entity to expand on
 * @param expands - The expands to transform to a query parameter
 * @param entityConstructor - Constructor type of the entity to expand on
 * @returns An object containing the query parameter or an empty object
 */
```

File: packages/core/src/odata/common/entity-deserializer.ts (hunk 3 above)
Comment (FrankEssenberger) [reply packages/core/src/odata/common/entity-deserializer.ts:32]: I will try to remove this. If we export it would be a breaking change.

**Pull request merged**